### PR TITLE
ceph rbd-mirror: configure entirely in operator

### PR DESCRIPTION
**Pull request by BlaineEXE**
In the form of mon, mgr, mds, and rgw, convert the rbd-mirror to be
configured completely from the operator. This is a straightforward
conversion with one functional addition: the rbd-mirror daemon stores
*no* data and has no default data dir, so the concept of a `Dataless`
daemon is here introduced.

This is the last one before the osds, which are expected to be hairy. I'll try to get them done in small chunks, one pod at a time.

Signed-off-by: Blaine Gardner <blaine.gardner@suse.com>

**Checklist:**
- [x] Documentation has been updated, if necessary.
- [x] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [x] Code generation (`make codegen`) has been run to update object specifications, if necessary.
- [x] Comments have been added or updated based on the standards set in [CONTRIBUTING.md](../blob/master/CONTRIBUTING.md#comments)

### CODE REVIEW

**Comment by BlaineEXE** [comment]
I noticed while working on this that there is no `UpdateDeploymentAndWait` for the rbd mirrors. That should be simple, so I think I will do it in another PR (on top of this one) starting Monday. I'll probably opt to do all of my upgrade testing on that PR than this one.

File: pkg/operator/ceph/cluster/rbd/mirror.go [hunk 1]
```diff
@@ -37,8 +38,9 @@ const (
 	appName = "rook-ceph-rbd-mirror"
 )
 
-// Cluster represents the Rook and environment configuration settings needed to set up rbd mirroring.
-type Mirroring struct {
+// ClusterConfig represents the Rook and environment configuration settings needed to set up rbd mirroring.
+type ClusterConfig struct {
```
Comment (leseb) [inline pkg/operator/ceph/cluster/rbd/mirror.go]: Although this makes perfect sense from a consistency point of view, that's not the original goal of the PR and not mentioned in your commit message.

File: pkg/operator/ceph/file/mds/config.go [hunk 2]
```diff
@@ -55,9 +55,9 @@ func (c *Cluster) generateKeyring(m *mdsConfig, deploymentUID types.UID) error {
 	err = c.context.Clientset.CoreV1().Secrets(c.fs.Namespace).Delete(m.ResourceName, &metav1.DeleteOptions{})
 	if err != nil {
 		if errors.IsNotFound(err) {
-			logger.Debugf("legacy mgr key %s is already removed", m.ResourceName)
+			logger.Debugf("legacy mds key %s is already removed", m.ResourceName)
```
Comment (leseb) [inline pkg/operator/ceph/file/mds/config.go]: Same here for the MDS, I think a separate commit is more appropriate.

File: pkg/operator/ceph/cluster/rbd/mirror.go [hunk 3]
```diff
@@ -70,22 +82,25 @@ func New(context *clusterd.Context, namespace, rookVersion string, cephVersion c
...
+			return fmt.Errorf("failed to generate keyring for %s. %+v", resourceName, err)
 		}
 
 		// Start the deployment
-		deployment := m.makeDeployment(resourceName, daemonName)
+		deployment := m.makeDeployment(daemonConf)
 		if _, err := m.context.Clientset.Apps().Deployments(m.Namespace).Create(deployment); err != nil {
 			if !errors.IsAlreadyExists(err) {
 				return fmt.Errorf("failed to create %s deployment. %+v", resourceName, err)
+				/* TODO: update rbd mirrors if already exist */
```
Comment (travisn) [inline pkg/operator/ceph/cluster/rbd/mirror.go]: This will be a separate PR? Definitely would be good to take care of this or else we won't get upgrades.

Comment (travisn) [reply pkg/operator/ceph/cluster/rbd/mirror.go]: Ah I see it's #2804
